metricflow/model/semantics/valid_join.py

```diff
@@ -0,0 +1,123 @@
+from __future__ import annotations
+
+from dataclasses import dataclass
+
+from metricflow.instances import DataSourceReference, DataSourceElementReference, InstanceSet
+from metricflow.model.objects.elements.identifier import IdentifierType
+from metricflow.protocols.semantics import DataSourceSemanticsAccessor
+from metricflow.references import IdentifierReference
+
+
+@dataclass(frozen=True)
+class DataSourceIdentifierJoinType:
+    """Describe a type of join between data sources where identifiers are of the listed types."""
+
+    left_identifier_type: IdentifierType
+    right_identifier_type: IdentifierType
+
+
+class DataSourceJoinValidator:
```

Comment:
Got any good ideas for adding explicit tests on the logic in this class? If not I'll write some test cases using the SCD model as a basis, since that change could be a little bit finicky.

Reply:
In retrospect, this should have a set. To unblock you though, seems best to merge and then handle later.